File: classic puzzle-medium/longest-increasing-subsequence.cpp
//https://www.codingame.com/training/medium/longest-increasing-subsequence
//LONGEST INCREASING SUBSEQUENCE
#include <bits/stdc++.h>

using namespace std;

int main()
{
    int n;
    cin >> n; cin.ignore();
    int arr[2570];
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }
    
    int m[2570]={0};
    for(int i=0;i<n;i++) m[i]=1;

    for(int i=0;i<n;i++){
        for(int j=i;j<n;j++){
            if(arr[j]>arr[i] && m[j]<m[i]+1)
                m[j]=m[i]+1;
        }
    }

    cout <<*max_element(m,m+n)<< endl;
}

File: classic puzzle-medium&above/longest-increasing-subsequence.rb
#https://www.codingame.com/training/medium/longest-increasing-subsequence
#LONGEST INCREASING SUBSEQUENCE
m=[]
n=gets.to_i.times do
    m.append(gets.to_i)
end

a=[1]*n
for i in (0...n)
    for j in (i...n)
        if m[j]>m[i] && a[j]<a[i]+1
            a[j]=a[i]+1
        end
    end
end

puts a.max
